### PR TITLE
ForkID and LWMA

### PR DESCRIPTION
**Pull request by Vutov**
Review only, its still not connecting to testnet

### CODE REVIEW

File: blockchain/difficulty.go [hunk 1]
```diff
@@ -224,6 +225,26 @@ func (b *BlockChain) calcNextRequiredDifficulty(lastNode *blockNode, newBlockTim
 		return b.chainParams.PowLimitBits, nil
 	}
 
+	config := b.chainParams.LWMA
+	if lastNode.height >= config.EnableHeight {
```
Comment (h4x3rotab) [inline blockchain/difficulty.go:229]: Did you consider the difficulty reduce window before the actual mainnet LWMA launch?
https://github.com/BTCGPU/BTCGPU/blob/76f46c4048718a49b52d724720e8620b5daac68f/src/pow.cpp#L52

File: chaincfg/params.go [hunk 2]
```diff
@@ -687,3 +742,8 @@ func init() {
 	mustRegister(&RegressionNetParams)
 	mustRegister(&SimNetParams)
 }
+
+func createPow(value string) *big.Int {
```
Comment (h4x3rotab) [inline chaincfg/params.go]: Better name it something like "powTargetFromString"

**Review by h4x3rotab — CHANGES_REQUESTED** [review]

File: chaincfg/params.go (hunk 2 above)
Comment (Vutov) [reply chaincfg/params.go]: Renamed

File: blockchain/difficulty.go (hunk 1 above)
Comment (Vutov) [reply blockchain/difficulty.go:229]: We can skip that check. We can use checkpoint right after the fork to verify the chain`s integrity up until that point and continue with 'normal' sync after. (or use an other checkpoint to save some sync time)

Comment (h4x3rotab) [reply blockchain/difficulty.go:229]: Can users disable the checkpoint to performa full verification on btgd? If not, it's reasonable to ignore.

Comment (Vutov) [reply blockchain/difficulty.go:229]: If they removed the code they can skip it, not sure if there is any other way - maybe there is.

Comment (h4x3rotab) [reply blockchain/difficulty.go:229]: Please double check if there's any config or command line flag to disable it. Of course changing of the code is not considered a feature.

Comment (Vutov) [reply blockchain/difficulty.go:229]: It seems checkpointing cannot be bypassed 
```// Ensure chain matches up to predetermined checkpoints.
	blockHash := header.BlockHash()
	if !b.verifyCheckpoint(blockHeight, &blockHash) {
		str := fmt.Sprintf("block at height %d does not match "+
			"checkpoint hash", blockHeight)
		return ruleError(ErrBadCheckpoint, str)
	}```

**Comment by Vutov** [comment]
Mainnet sync complete to chain tip

File: blockchain/process.go [hunk 3]
```diff
@@ -143,7 +143,8 @@ func (b *BlockChain) ProcessBlock(block *btcutil.Block, flags BehaviorFlags) (bo
 	b.chainLock.Lock()
 	defer b.chainLock.Unlock()
 
-	fastAdd := flags&BFFastAdd == BFFastAdd
+	// Not needed
```
Comment (h4x3rotab) [inline blockchain/process.go]: Delete the lines

File: blockchain/process.go [hunk 4]
```diff
@@ -190,24 +191,28 @@ func (b *BlockChain) ProcessBlock(block *btcutil.Block, flags BehaviorFlags) (bo
...
-			if currentTarget.Cmp(requiredTarget) > 0 {
-				str := fmt.Sprintf("block target difficulty of %064x "+
-					"is too low when compared to the previous "+
-					"checkpoint", currentTarget)
-				return false, false, ruleError(ErrDifficultyTooLow, str)
-			}
-		}
+		// BTG uses LWMA, having any kind of prediction based on checkpoint is ambiguous,
+		// next target can be calculated based on last block - difficulty.go
+		// if !fastAdd {
```
Comment (h4x3rotab) [inline blockchain/process.go]: Delete the unused lines and keep the above comments

File: chaincfg/params.go [hunk 5]
```diff
@@ -425,8 +465,11 @@ var TestNet3Params = Params{
 
 	// Checkpoints ordered from oldest to newest.
 	Checkpoints: []Checkpoint{
-		{0, newHashFromStr("00000000e0781ebe24b91eedc293adfea2f557b53ec379e78959de3853e6f9f6")},
+		{1, newHashFromStr("0003ce866d462b6ef4d229021c603644ae9f405f4952ee5f6c03ce13b37440d9")},
+		{14300, newHashFromStr("0004171c096c6a77949455d57399f2c9eeadec83bbe4307cf7db00b2369da39a")},
+		{44000, newHashFromStr("0000815745f2fcf4a0b2baf8a04e0fdcf3a08bcea8972c1f1fdf3e9f995f4795")},
 	},
+	// TODO : Rejected block 00004324eb108c302e6ecb10beaad916c2a73f3e40821d3654bb366c8c9faf1a block height 32, bypassed by checkointing
```
Comment (h4x3rotab) [inline chaincfg/params.go]: This block was resolved, right?

**Review by h4x3rotab — CHANGES_REQUESTED** [review]
Come on, it's almost done!

File: chaincfg/params.go (hunk 5 above)
Comment (Vutov) [reply chaincfg/params.go]: Yes, will remove the TODO.

File: blockchain/process.go (hunk 3 above)
Comment (Vutov) [reply blockchain/process.go]: Will do

File: blockchain/process.go (hunk 4 above)
Comment (Vutov) [reply blockchain/process.go]: Will do

**Review by h4x3rotab — APPROVED** [review]
Good job

**Pull request merged**